### PR TITLE
fix the amm 

### PR DESCRIPTION
**Pull request by riverman123**
1 use netQuoteToken to calculate the netBaseToken
2 add price check to ensure the price won't be changed before and after removing liquidity
3  add config value that the left quoteToken can cover the netQuoteToken liquidated.
4 only router can add and remove liquidity
5  rebase use the origin way that the price and be the same after rebase rather than a step rise.

### CODE REVIEW

File: contracts/mocks/MockMargin.sol [hunk 1]
```diff
@@ -27,8 +27,13 @@ contract MockMargin  {
         config = config_ ;
     }
 
-    function netPosition() external view returns (int256 netBasePosition){
-        return 2 * 10**18;
+    function netPosition() external view returns (int256 ){
```
Comment (aupiff) [inline contracts/mocks/MockMargin.sol:30]: this function is created automatically, no need to write it.

File: contracts/mocks/MockMargin.sol [hunk 2]
```diff
@@ -27,8 +27,13 @@ contract MockMargin  {
...
+        return netpostion;
+    }
+    function setNetPosition(int256 newNetpositon ) external  returns (int256 ){
+        // return 200 * 10 **6;
```
Comment (aupiff) [inline contracts/mocks/MockMargin.sol:34]: this comment is strange. remove it?

File: contracts/mocks/MockMargin.sol [hunk 3]
```diff
@@ -50,4 +55,12 @@ contract MockMargin  {
     ) external returns (uint256[2] memory amounts) {
       IAmm(amm).swap(trader, inputToken, outputToken, inputAmount, outputAmount);
     }
+
+       function withdraw(
```
Comment (aupiff) [inline contracts/mocks/MockMargin.sol:59]: does this function do anything? did you mean to remove it?

File: contracts/core/Margin.sol [hunk 4]
```diff
@@ -562,8 +562,8 @@ contract Margin is IMargin, IVault, Reentrant {
     }
 
     function netPosition() external view override returns (int256) {
-        require(totalQuoteLong < type(uint128).max, "Margin.netPosition: OVERFLOW");
-        return int256(totalQuoteLong).subU(totalQuoteShort);
+        require(totalQuoteShort < type(uint128).max, "Margin.netPosition: OVERFLOW"); 
+        return int256(totalQuoteShort).subU(totalQuoteLong);
```
Comment (aupiff) [inline contracts/core/Margin.sol:566]: why did this just change sign? was there some error before?

File: contracts/mocks/MockMargin.sol (hunk 3 above)
Comment (riverman123) [reply contracts/mocks/MockMargin.sol:59]: just for test pass

File: contracts/core/Amm.sol [hunk 5]
```diff
@@ -123,30 +138,83 @@ contract Amm is IAmm, LiquidityERC20, Reentrant {
...
+            quoteTokenOfNetPosition.abs() * 100 < (_quoteReserve - quoteAmount) * lpWithdrawThreshold,
+            "Amm.burn: TOO_LARGE_LIQUIDITY_WITHDRAW"
+        );
+
+   
+        require(
+            (_baseReserve - baseAmount) * _quoteReserve * 999 <= (_quoteReserve - quoteAmount) * _baseReserve * 1000,
+            "Amm.burn: PRICE_BEFORE_AND_AFTER_MUST_BE_THE_SAME"
+        );
+        require(
```
Comment (aupiff) [inline contracts/core/Amm.sol:176]: there's no reason to have this be 1000 & 999 in one check and 1001 and 1000 in the other, right?

**Pull request merged**